File: py/tool/follow_flow.py
"""re/py/tool/follow_flow.py — LUỒNG FOLLOW no-phone: mỗi account login(2135 full)→follow target→verify STUCK/shadow.
   Port re/tests/follow_flow.mjs. Spawn worker.py/tiến trình (mỗi account 1 PROXY_URL — net.py fix proxy theo process).
  python follow_flow.py <target> [type=1|0] [--acc accounts.txt] [--proxy proxies.txt] [--conc N]
    accounts.txt: mỗi dòng "user|pass|email|mailpass"
    proxies.txt : mỗi dòng "ip:port:user:pass"
"""
import os
import re
import sys
import subprocess
from concurrent.futures import ThreadPoolExecutor

HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, os.path.dirname(HERE))
try:
    sys.stdout.reconfigure(encoding='utf-8', errors='replace')
except Exception:
    pass

WORKER = os.path.join(HERE, 'worker.py')


def flag(name, default=None):
    a = sys.argv
    return a[a.index(name) + 1] if name in a and a.index(name) + 1 < len(a) else default


def to_proxy(line):
    if '://' in line:
        return line
    a = line.split(':')
    if len(a) == 4:
        return f'http://{a[2]}:{a[3]}@{a[0]}:{a[1]}'
    if len(a) == 2:
        return f'http://{a[0]}:{a[1]}'
    return line


def read_lines(path):
    if not path or not os.path.exists(path):
        return []
    with open(path, encoding='utf-8') as f:
        return [x.strip() for x in f if x.strip() and not x.strip().startswith('#')]


def run_one(idx, total, account, proxy, target, ftype):
    user = account.split('|')[0]
    env = {**os.environ, 'ACCOUNT': account, 'FOLLOW': target, 'FOLLOW_TYPE': ftype, 'NO_PAUSE': '1'}
    if proxy:
        env['PROXY_URL'] = proxy
    try:
        p = subprocess.run([sys.executable, WORKER, account], env=env, capture_output=True,
                           text=True, timeout=280, encoding='utf-8', errors='replace')
        out = (p.stdout or '') + (p.stderr or '')
    except subprocess.TimeoutExpired:
        out = '(timeout)'
    session = 'THÔNG TIN TÀI KHOẢN' in out
    uid = (re.search(r'uid\s*:\s*(\d+)', out) or [None, None])[1]
    m = re.search(r'FOLLOW @\S+: sc=(-?\d+) follow_status \d+→(\d+) → (.+)', out)
    if not session:
        ec = (re.search(r'ec=(\d+)', out) or [None, None])[1]
        status = 'ec7(throttle/IP)' if (ec == '7' or 'Maximum' in out) else f'login-fail({ec or "?"})'
        verdict = 'nologin'
    elif m:
        after = m.group(2)
        verdict = 'stuck' if after == '1' else ('shadow' if m.group(1) == '0' else 'fail')
        status = {'stuck': '✅ STUCK', 'shadow': '⚠️ shadow-drop'}.get(verdict, f'follow-{verdict}({m.group(1)})')
    else:
        status, verdict = 'session-ok/follow-?', 'unknown'
    proxy_show = proxy.split('@')[-1] if proxy else 'direct'
    print(f"  [{idx + 1}/{total}] {user} | {proxy_show} → {'SESSION ' + (uid or '') if session else 'NO-SESSION'} | follow: {status}", flush=True)
    return {'user': user, 'uid': uid, 'session': session, 'verdict': verdict, 'status': status}


def main():
    if len(sys.argv) < 2 or sys.argv[1].startswith('--'):
        print('cần <target>. VD: python follow_flow.py idmahg 1 --acc accounts.txt --proxy proxies.txt')
        return 2
    target = sys.argv[1].lstrip('@')
    ftype = sys.argv[2] if len(sys.argv) > 2 and sys.argv[2] in ('0', '1') else '1'
    accounts = read_lines(flag('--acc', os.path.join(HERE, 'accounts.txt')))
    proxies = [to_proxy(x) for x in read_lines(flag('--proxy', os.path.join(HERE, 'proxies.txt')))]
    conc = int(flag('--conc', '2'))
    if not accounts:
        print('accounts.txt trống (user|pass|email|mailpass mỗi dòng)')
        return 2
    print(f"[follow_flow] {len(accounts)} account → follow @{target} (type={ftype}) | {len(proxies)} proxy | conc={conc}")

    results = [None] * len(accounts)
    with ThreadPoolExecutor(max_workers=conc) as ex:
        futs = {ex.submit(run_one, i, len(accounts), a, (proxies[i % len(proxies)] if proxies else ''),
                          target, ftype): i for i, a in enumerate(accounts)}
        for fut in futs:
            i = futs[fut]
            results[i] = fut.result()

    stuck = sum(1 for r in results if r and r['verdict'] == 'stuck')
    shadow = sum(1 for r in results if r and r['verdict'] == 'shadow')
    nologin = sum(1 for r in results if r and not r['session'])
    print(f"\n===== TỔNG KẾT follow @{target} =====")
    print(f"✅ STUCK (follow thật): {stuck}/{len(results)}  |  ⚠️ shadow-drop: {shadow}  |  ❌ no-session: {nologin}")
    for r in results:
        if r:
            print(f"  {r['user']} → {r['status']}" + (f" (uid {r['uid']})" if r['uid'] else ''))
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: tests/follow_flow.mjs
// follow_flow.mjs — LUỒNG FOLLOW no-phone hoàn chỉnh.
//   Mỗi account: forge device → login (2135 full: send_code→code_login→2135→authenticate→re-login #17) → session
//                → resolve sec_uid target (aweme search) → follow → VERIFY stick (re-search follow_status).
//   Chạy dưới lớp t_full_session.mjs (đã proven login+follow+verify). Mỗi account 1 proxy riêng, xoay vòng.
//
//   node re/tests/follow_flow.mjs <target_uniqueId> [type=1|0] [--acc <file>] [--proxy <file>] [--conc N]
//     accounts file: mỗi dòng "user|pass|email|mailpass"
//     proxies file : mỗi dòng "ip:port:user:pass"  (rỗng = đi thẳng, KHÔNG khuyến nghị)
//   env: OMO_API_KEY (captcha send_code), RE_VER (mặc định 45.7.3)
import { spawn } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const DIR = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(DIR, '..', '..');
const args = process.argv.slice(2);
const target = (args[0] || '').replace(/^@/, '');
if (!target) { console.error('cần <target_uniqueId>. VD: node re/tests/follow_flow.mjs idmahg 1 --acc accounts.txt --proxy proxies.txt'); process.exit(1); }
const type = /^[01]$/.test(args[1]) ? args[1] : '1';
const flag = (n, d) => { const i = args.indexOf(n); return i >= 0 && args[i + 1] ? args[i + 1] : d; };
const accFile = flag('--acc', path.join(DIR, 'accounts.txt'));
const proxyFile = flag('--proxy', path.join(DIR, 'proxies.txt'));
const conc = parseInt(flag('--conc', '2'), 10);
const OMO = process.env.OMO_API_KEY || '';

if (!fs.existsSync(accFile)) { console.error('không thấy accounts file: ' + accFile + '  (mỗi dòng user|pass|email|mailpass)'); process.exit(1); }
const accounts = fs.readFileSync(accFile, 'utf8').trim().split('\n').map((l) => l.trim()).filter(Boolean);
const proxies = fs.existsSync(proxyFile)
  ? fs.readFileSync(proxyFile, 'utf8').trim().split('\n').map((l) => l.trim()).filter(Boolean).map((l) => { const [ip, port, u, p] = l.split(':'); return u ? `http://${u}:${p}@${ip}:${port}` : `http://${ip}:${port}`; })
  : [];

console.log(`[follow_flow] ${accounts.length} account → follow @${target} (type=${type}) | ${proxies.length} proxy | conc=${conc}`);

function runOne(account, idx) {
  return new Promise((resolve) => {
    const proxy = proxies.length ? proxies[idx % proxies.length] : '';
    const user = account.split('|')[0];
    const env = { ...process.env, FOLLOW: target, FOLLOW_TYPE: type, RE_VER: process.env.RE_VER || '45.7.3', NO_COMPILE: '1', OMO_API_KEY: OMO };
    if (proxy) env.PROXY = proxy;
    const p = spawn(process.execPath, [path.join(DIR, 't_full_session.mjs'), account], { env, cwd: ROOT });
    let out = '';
    p.stdout.on('data', (d) => { out += d; });
    p.stderr.on('data', () => {});
    const to = setTimeout(() => { try { p.kill(); } catch {} }, 260000);
    p.on('close', () => {
      clearTimeout(to);
      const session = /🎉🎉🎉 SESSION!/.test(out);
      const uid = (out.match(/user_id=(\d+)/) || [])[1];
      const followSc = (out.match(/\[F2\][^\n]*status_code=(-?\d+)/) || [])[1];
      const after = (out.match(/follow_status\(after\)=(\d+)/) || [])[1];
      const ec = (out.match(/user\/login=\s*(\d+)/) || [])[1];
      const stuck = after === '1';
      const status = !session ? (ec === '7' ? 'ec7(throttle/IP)' : 'login-fail(' + (ec || '?') + ')') : (followSc === '0' ? (stuck ? '✅ STUCK' : '⚠️ shadow-drop') : 'follow-fail(' + (followSc || '?') + ')');
      console.log(`  [${idx + 1}/${accounts.length}] ${user} | proxy ${proxy ? proxy.split('@')[1] : 'direct'} → ${session ? 'SESSION ' + (uid || '') : 'NO-SESSION'} | follow: ${status}`);
      resolve({ user, uid, session, followSc, after, stuck, status });
    });
  });
}

(async () => {
  const results = [];
  for (let i = 0; i < accounts.length; i += conc) {
    const batch = accounts.slice(i, i + conc).map((a, j) => runOne(a, i + j));
    results.push(...await Promise.all(batch));
  }
  const stuck = results.filter((r) => r.stuck).length;
  const shadow = results.filter((r) => r.session && r.followSc === '0' && !r.stuck).length;
  const nologin = results.filter((r) => !r.session).length;
  console.log('\n===== TỔNG KẾT follow @' + target + ' =====');
  console.log(`✅ STUCK (follow thật): ${stuck}/${results.length}  |  ⚠️ shadow-drop: ${shadow}  |  ❌ no-session: ${nologin}`);
  results.forEach((r) => console.log(`  ${r.user} → ${r.status}${r.uid ? ' (uid ' + r.uid + ')' : ''}`));
})();
